service/topology_coordinator.cc

```diff
@@ -2066,6 +2079,137 @@ future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard gua
     co_return true;
 }
 
+future<bool> topology_coordinator::do_upgrade_step(group0_guard guard) {
+    switch (_topo_sm._topology.ustate) {
+    case topology::upgrade_state::not_upgraded:
+        on_internal_error(rtlogger, std::make_exception_ptr(std::runtime_error(
+                "topology_coordinator was started even though upgrade to raft topology was not requested yet")));
+
+    case topology::upgrade_state::build_coordinator_state:
+        co_await build_coordinator_state(std::move(guard));
+        co_return true;
+
+    case topology::upgrade_state::final_global_barrier: {
```

Comment:
Why do we need this final global barrier? What is it waits for?

Reply:
> given that the administrator has to perform a "global barrier" themselves by checking the upgrade state on all node.

The coordinator may automatically start some work after this barrier, e.g. suppose we add further "upgrade" steps for auth, service levels, tablets...

But maybe those steps will start with a barrier

Although this barrier definitely doesn't hurt. It's also needed if we want the POST to block

Reply:
> > given that the administrator has to perform a "global barrier" themselves by checking the upgrade state on all node.
> 
> The coordinator may automatically start some work after this barrier, e.g. suppose we add further "upgrade" steps for auth, service levels, tablets...
> 
> But maybe those steps will start with a barrier

Everything that is driven by the coordinator goes through raft cmd rpc which runs barrier to see the latest state.
Explicit one is not needed usually.

> 
> Although this barrier definitely doesn't hurt. It's also needed if we want the POST to block

We will need another barrier after we move to "done" state then :) Lets not do barriers we do not need (yes they are harmless, but just complicate code unnecessary).